crates/goose/src/systems/non_developer.rs

```diff
@@ -0,0 +1,971 @@
+                        Automate tasks in video or music production tools like Final Cut Pro or GarageBand.
+                    Data Processing
+                        Process text files or other types of documents.
+                        Extract or format data from files or apps.
+                        Interact with spreadsheets (e.g., Numbers or Excel).
+                    Integration with Other Scripts
+                        Execute shell scripts, Ruby scripts, or other automation scripts.
+                        Combine workflows across scripting languages.
+                    Complex Workflows
+                        Automate multi-step tasks involving multiple apps or system features.
+                        Create scheduled tasks using Calendar or other scheduling apps.
+
+            web_search
+              - Search the web using DuckDuckGo's API for general topics or keywords
+
+            cache
+              - Manage your cached files
+              - List, view, delete files
+              - Clear all cached data
+
+            The system automatically manages:
+            - Cache directory: {cache_dir}
+            - File organization and cleanup
+            "#,
+            cache_dir = cache_dir.display()
+        };
+
+        Self {
+            tools: vec![
+                web_search_tool,
+                web_scrape_tool,
+                data_process_tool,
+                quick_script_tool,
+                cache_tool,
+            ],
+            cache_dir,
+            active_resources: Mutex::new(HashMap::new()),
+            http_client: Client::builder().user_agent("Goose/1.0").build().unwrap(),
+            instructions,
+        }
+    }
+
+    // Helper function to generate a cache file path
+    fn get_cache_path(&self, prefix: &str, extension: &str) -> PathBuf {
+        let timestamp = chrono::Local::now().format("%Y%m%d_%H%M%S");
+        self.cache_dir
+            .join(format!("{}_{}.{}", prefix, timestamp, extension))
+    }
+
+    // Helper function to save content to cache
+    async fn save_to_cache(
+        &self,
+        content: &[u8],
+        prefix: &str,
+        extension: &str,
+    ) -> AgentResult<PathBuf> {
+        let cache_path = self.get_cache_path(prefix, extension);
+        fs::write(&cache_path, content)
+            .map_err(|e| AgentError::ExecutionError(format!("Failed to write to cache: {}", e)))?;
+        Ok(cache_path)
+    }
+
+    // Implement web_scrape tool functionality
+    async fn web_search(&self, params: Value) -> AgentResult<Vec<Content>> {
+        let query = params
+            .get("query")
+            .and_then(|v| v.as_str())
+            .ok_or_else(|| AgentError::InvalidParameters("Missing 'query' parameter".into()))?;
+
+        // Create the DuckDuckGo API URL
+        let url = format!(
+            "https://api.duckduckgo.com/?q={}&format=json&pretty=1",
```

Comment:
I was getting some weird results from their api. couldn't find any official pages on the api, but some older [posts](https://stackoverflow.com/questions/37012469/duckduckgo-api-getting-search-results) on stack overflow suggest that it's not great

Reply:
yeah it really only works with topics - like single words, eg try this: https://api.duckduckgo.com/?q=bridgestone&format=json&pretty=1 - good for proper nouns which may not be in training set?

Reply:
I'll tighten it up to make sure only used for proper nouns (where it does come in handy)